codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +321,509 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
         for (int i = 0; i < channelsToFireChildReadComplete.size(); i++) {
-            Http2StreamChannel childChannel = channelsToFireChildReadComplete.get(i);
+            DefaultHttp2StreamChannel childChannel = channelsToFireChildReadComplete.get(i);
             // Clear early in case fireChildReadComplete() causes it to need to be re-processed
             childChannel.inStreamsToFireChildReadComplete = false;
             childChannel.fireChildReadComplete();
         }
         channelsToFireChildReadComplete.clear();
     }
 
-    ChannelFuture createStreamChannel(Channel parentChannel, EventLoopGroup group, ChannelHandler handler,
-                                              Map<ChannelOption<?>, Object> options,
-                                              Map<AttributeKey<?>, Object> attrs,
-                                              int streamId) {
-        final Http2StreamChannel channel = new Http2StreamChannel(parentChannel);
-        if (isStreamIdValid(streamId)) {
-            assert !isOutboundStream(server, streamId);
-            assert ctx.channel().eventLoop().inEventLoop();
-            channel.streamId(streamId);
-        }
-        channel.pipeline().addLast(handler);
-
-        initOpts(channel, options);
-        initAttrs(channel, attrs);
-
-        ChannelFuture future = group.register(channel);
-        // Handle any errors that occurred on the local thread while registering. Even though
-        // failures can happen after this point, they will be handled by the channel by closing the
-        // channel.
-        if (future.cause() != null) {
-            if (channel.isRegistered()) {
-                channel.close();
-            } else {
-                channel.unsafe().closeForcibly();
+    private static final class DefaultHttp2StreamChannel extends AbstractChannel implements Http2StreamChannel {
+
+        private final ChannelHandlerContext ctx;
+        private final ChannelFutureListener firstFrameWriteListener;
+
+        /** {@code true} after the first HEADERS frame has been written **/
+        private boolean firstFrameWritten;
+
+        /** {@code true} if a close without an error was initiated **/
+        private boolean streamClosedWithoutError;
+
+        /** {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}. **/
+        boolean inStreamsToFireChildReadComplete;
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<DefaultHttp2StreamChannel> OUTBOUND_FLOW_CONTROL_WINDOW_UPDATER;
+
+        /**
+         * Used by subclasses to queue a close channel within the read queue. When read, it will close
+         * the channel (using Unsafe) instead of notifying handlers of the message with {@code
+         * channelRead()}. Additional inbound messages must not arrive after this one.
+         */
+        protected static final Object CLOSE_MESSAGE = new Object();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "protected" modifier. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2156)